core/src/smartcontracts/isi/permissions/mod.rs

```diff
@@ -2,105 +2,232 @@
+    }
+
+    /// Check if verdict is [`Deny`](ValidatorVerdict::Deny)
+    #[inline]
+    pub fn is_deny(&self) -> bool {
+        matches!(self, ValidatorVerdict::Deny(_))
+    }
 
-    use super::{Decode, Encode, IntoSchema, ValidatorType};
-    use crate::smartcontracts::Mismatch;
+    /// Check if verdict is [`Skip`](ValidatorVerdict::Skip)
+    #[inline]
+    pub fn is_skip(&self) -> bool {
+        matches!(self, ValidatorVerdict::Skip)
+    }
+
+    /// Compare `self` with `other` and return the least permissive one
+    ///
+    /// Returns `self` if both are equal
+    #[must_use]
+    #[inline]
+    pub fn least_permissive(self, other: Self) -> Self {
+        std::cmp::min(self, other)
+    }
+
+    /// Similar to [`least_permissive`](Self::least_permissive)
+    /// but won't compute `f` if `self` is [`Deny`](ValidatorVerdict::Deny)
+    #[must_use]
+    pub fn least_permissive_with(self, f: impl FnOnce() -> Self) -> Self {
+        if let Self::Deny(_) = &self {
+            self
+        } else {
+            self.least_permissive(f())
+        }
+    }
 
-    /// Wrong validator expectation error
+    /// Compare `self` with `other` and return the most permissive one
     ///
-    /// I.e. used when user tries to validate [`QueryBox`](super::QueryBox) with
-    /// [`IsAllowedBoxed`](super::IsAllowedBoxed) containing
-    /// [`IsAllowedBoxed::Instruction`](super::IsAllowedBoxed::Instruction) variant
-    pub type ValidatorTypeMismatch = Mismatch<ValidatorType>;
-
-    /// Reason for prohibiting the execution of the particular instruction.
-    #[derive(Debug, Clone, thiserror::Error, Decode, Encode, IntoSchema)]
-    #[allow(variant_size_differences)]
-    pub enum DenialReason {
-        /// [`ValidatorTypeMismatch`] variant
-        #[error("Wrong validator type: {0}")]
-        ValidatorTypeMismatch(#[from] ValidatorTypeMismatch),
-        /// Variant for custom error
-        #[error("{0}")]
-        Custom(String),
-        /// Variant used when at least one [`Validator`](super::IsAllowed) should be provided
-        #[error("No validators provided")]
-        NoValidatorsProvided,
+    /// Returns `self` if both are equal
+    #[must_use]
+    #[inline]
+    pub fn most_permissive(self, other: Self) -> Self {
+        std::cmp::max(self, other)
     }
 
-    impl From<String> for DenialReason {
-        fn from(s: String) -> Self {
-            Self::Custom(s)
+    /// Similar to [`most_permissive`](Self::most_permissive)
+    /// but won't compute `f` if `self` is [`Allow`](ValidatorVerdict::Allow)
+    #[must_use]
+    pub fn most_permissive_with(self, f: impl FnOnce() -> Self) -> Self {
+        if let Self::Allow = &self {
+            self
+        } else {
+            self.most_permissive(f())
         }
     }
 
-    impl FromStr for DenialReason {
-        type Err = Infallible;
+    // Not implemented as some trait implementation cause it should stay private
+    fn as_u8(&self) -> u8 {
```

Comment:
I'd consider adding a standard trait as well. For example `Deref` might be a candidate.

Reply:
Okay, I'll remove `as_u8()` then